File: J9Ch/src/S_6_Dynamic_Problem_II/Optional_11/_89_k_Sum_Hard.py

class Solution:

    def kSum(self, A, k, target):
        ans = [[[0 for i in range(target + 1)] for j in range(k + 1)] for K in range(len(A) + 1)]

        ans[0][0][0] = 1
        for I in range(len(A)):
            item = A[I]
            for J in range(target + 1):
                for K in range(k + 1):
                    tk = k - K
                    tj = target - J
                    ans[I + 1][tk][tj] = ans[I][tk][tj]
                    if tk - 1 >= 0 and tj - item >= 0:
                        ans[I + 1][tk][tj] += ans[I][tk - 1][tj - item]
        return ans[len(A)][k][target]

File: J9Ch/src/S_6_DP_II/Optional_11/python_cpp/_89_k_Sum_Hard.cpp

class Solution {
public:
    int ans[105][1005];
    int kSum(vector<int> A, int K, int target) {
        ans[0][0] = 1;
        for(int i = 0; i < A.size(); ++i)
            for(int j = K; j > 0; j--)
                for(int k = target; k >= A[i]; k--) {
                    ans[j][k] += ans[j - 1][k - A[i]];
                    if(ans[j][k] > 2147483647)
                        return -1;
                }
        return ans[K][target];
    }
};